wrench/src/main.rs

```diff
@@ -75,31 +86,161 @@ impl ThingKind {
     }
 }
 
+#[cfg(feature = "headless")]
+pub struct HeadlessContext {
+    width: u32,
+    height: u32,
+    _context: osmesa_sys::OSMesaContext,
+    _buffer: Vec<u32>,
+}
+
+#[cfg(not(feature = "headless"))]
+pub struct HeadlessContext {
+    width: u32,
+    height: u32,
+}
+
+impl HeadlessContext {
+    #[cfg(feature = "headless")]
+    fn new(width: u32, height: u32) -> HeadlessContext {
+        let mut attribs = Vec::new();
+
+        attribs.push(osmesa_sys::OSMESA_PROFILE);
+        attribs.push(osmesa_sys::OSMESA_CORE_PROFILE);
+        attribs.push(osmesa_sys::OSMESA_CONTEXT_MAJOR_VERSION);
+        attribs.push(3);
+        attribs.push(osmesa_sys::OSMESA_CONTEXT_MINOR_VERSION);
+        attribs.push(3);
+        attribs.push(0);
+
+        let context = unsafe {
+            osmesa_sys::OSMesaCreateContextAttribs(attribs.as_ptr(), ptr::null_mut())
+        };
+
+        assert!(!context.is_null());
+
+        let mut buffer = vec![0; (width * height) as usize];
+
+        unsafe {
+            let ret = osmesa_sys::OSMesaMakeCurrent(context,
+                                                    buffer.as_mut_ptr() as *mut _,
+                                                    gl::UNSIGNED_BYTE,
+                                                    width as i32,
+                                                    height as i32);
+            assert!(ret != 0);
+        };
+
+        HeadlessContext {
+            width: width,
+            height: height,
+            _context: context,
+            _buffer: buffer,
+        }
+    }
+
+    #[cfg(not(feature = "headless"))]
+    fn new(width: u32, height: u32) -> HeadlessContext {
+        HeadlessContext {
+            width: width,
+            height: height,
+        }
+    }
+
+    #[cfg(feature = "headless")]
+    fn get_proc_address(s: &str) -> *const c_void {
+        let c_str = CString::new(s).expect("Unable to create CString");
+        unsafe {
+            mem::transmute(osmesa_sys::OSMesaGetProcAddress(c_str.as_ptr()))
+        }
+    }
+
+    #[cfg(not(feature = "headless"))]
+    fn get_proc_address(_: &str) -> *const c_void {
+        ptr::null() as *const _
+    }
+}
+
+pub enum WindowWrapper {
+    Window(glutin::Window),
+    Headless(HeadlessContext),
+}
+
+pub struct HeadlessEventIterater;
+
+impl WindowWrapper {
+    fn init(&self) {
+        match *self {
+            WindowWrapper::Window(ref window) => {
+                unsafe {
+                    window.make_current().ok();
```

Comment:
I don't think it makes sense to ignore the returned `Err` here, we better just do `unwrap` or `expect` instead.